controllers/effective_dnspolicies_reconciler.go

```diff
@@ -0,0 +1,290 @@
+					rLogger.V(1).Info("listener hostname has changed, deleting record for listener")
+					r.deleteRecord(ctx, existingRecordObj)
+					//Break to allow it to try the creation of the desired record
+					break
+				}
+
+				if reflect.DeepEqual(existingRecord.Spec, desiredRecord.Spec) &&
+					reflect.DeepEqual(existingRecord.OwnerReferences, desiredRecord.OwnerReferences) {
+					rLogger.V(1).Info("dns record is up to date, nothing to do")
+					continue
+				}
+
+				rLogger.V(1).Info("updating DNS record for listener")
+				if _, uErr := resource.Update(ctx, obj, metav1.UpdateOptions{}); uErr != nil {
+					rLogger.Error(err, "unable to update dns record")
+				}
+				continue
+			}
+
+			if !hasAttachedRoute {
+				lLogger.V(1).Info("listener has no attached routes, skipping record create for listener")
+				continue
+			}
+
+			if len(desiredRecord.Spec.Endpoints) == 0 {
+				lLogger.V(1).Info("record for listener has no addresses, skipping record create for listener")
+				continue
+			}
+
+			//Create
+			lLogger.V(1).Info("creating DNS record for listener")
+			if _, cErr := resource.Create(ctx, obj, metav1.CreateOptions{}); cErr != nil && !apierrors.IsAlreadyExists(cErr) {
+				lLogger.Error(err, "unable to create dns record")
+			}
+		}
+
+		if !gatewayHasAddresses {
+			pLogger.V(1).Info("gateway has no addresses")
+			policyErrors[policy.GetLocator()] = ErrNoAddresses
+		} else if !gatewayHasAttachedRoutes {
+			pLogger.V(1).Info("gateway has no attached routes")
+			policyErrors[policy.GetLocator()] = ErrNoRoutes
+		}
+	}
+
+	state.Store(StateDNSPolicyErrorsKey, policyErrors)
+
+	return r.deleteOrphanDNSRecords(controller.LoggerIntoContext(ctx, logger), topology)
+}
+
+// listenersForPolicy returns an array of listeners that are targeted by the given policy.
+// If the target is a Listener a single element array containing that listener is returned.
+// If the target is a Gateway all listeners that do not have a DNS policy explicitly attached are returned.
+func (r *EffectiveDNSPoliciesReconciler) listenersForPolicy(_ context.Context, topology *machinery.Topology, policy machinery.Policy, policyTypeFilterFunc dnsPolicyTypeFilter) []*machinery.Listener {
+	return lo.Flatten(lo.FilterMap(topology.Targetables().Items(), func(t machinery.Targetable, _ int) ([]*machinery.Listener, bool) {
+		pTarget := lo.ContainsBy(t.Policies(), func(item machinery.Policy) bool {
+			return item.GetLocator() == policy.GetLocator()
+		})
+		if pTarget {
+			if l, ok := t.(*machinery.Listener); ok {
+				return []*machinery.Listener{l}, true
+			}
+			if g, ok := t.(*machinery.Gateway); ok {
+				listeners := lo.FilterMap(topology.Targetables().Children(g), func(t machinery.Targetable, _ int) (*machinery.Listener, bool) {
+					l, lok := t.(*machinery.Listener)
+					lPolicies := lo.FilterMap(l.Policies(), policyTypeFilterFunc)
+					return l, lok && len(lPolicies) == 0
+				})
+				return listeners, true
+			}
+		}
+		return nil, false
+	}))
+}
```

Comment:
Think we can use `topology.Targetables().Children(policy)` here to get the `Targetables` of a `Policy` which would make this a bit simpler 🤔 
```suggestion
func (r *EffectiveDNSPoliciesReconciler) listenersForPolicy(_ context.Context, topology *machinery.Topology, policy machinery.Policy, policyTypeFilterFunc dnsPolicyTypeFilter) []*machinery.Listener {
	return lo.Flatten(lo.FilterMap(topology.Targetables().Children(policy), func(t machinery.Targetable, _ int) ([]*machinery.Listener, bool) {
		if l, ok := t.(*machinery.Listener); ok {
			return []*machinery.Listener{l}, true
		}
		if g, ok := t.(*machinery.Gateway); ok {
			listeners := lo.FilterMap(topology.Targetables().Children(g), func(t machinery.Targetable, _ int) (*machinery.Listener, bool) {
				l, lok := t.(*machinery.Listener)
				lPolicies := lo.FilterMap(l.Policies(), policyTypeFilterFunc)
				return l, lok && len(lPolicies) == 0
			})
			return listeners, true
		}
		
		return nil, false
	}))
```

Reply:
Makes sense :+1: